pkg/runtime/grpc/k8s-portfwd-dialer.go

```
@@ -0,0 +1,157 @@
+func NewK8SPortFwdConn(ctx context.Context, pod target, timeout time.Duration) (net.Conn, error) {
+	conn := &k8sPortFwdDialer{}
+
+	config, err := utils.KubernetesConfigFlags.ToRESTConfig()
+	if err != nil {
+		return nil, fmt.Errorf("creating RESTConfig: %w", err)
+	}
+
+	// set GroupVersion and NegotiatedSerializer for RESTClient
+	factory.SetKubernetesDefaults(config)
+
+	conn.podName = pod.addressOrPod
+
+	config.Timeout = timeout
+
+	transport, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		return nil, fmt.Errorf("creating roundtripper: %w", err)
+	}
+
+	targetURL, err := url.Parse(config.Host)
+	if err != nil {
+		return nil, fmt.Errorf("parsing restConfig.Host: %w", err)
+	}
+
+	targetURL.Path = path.Join(
+		"api", "v1",
+		"namespaces", "gadget",
+		"pods", conn.podName,
+		"portforward",
+	)
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, targetURL)
+
+	newConn, _, err := dialer.Dial(portforward.PortForwardProtocolV1Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// create error stream
+	headers := http.Header{}
+	headers.Set(v1.StreamType, v1.StreamTypeError)
+	headers.Set(v1.PortHeader, fmt.Sprintf("%d", api.GadgetServicePort))
+	headers.Set(v1.PortForwardRequestIDHeader, strconv.Itoa(1))
+	errorStream, err := newConn.CreateStream(headers)
+	if err != nil {
+		newConn.Close()
+		return nil, fmt.Errorf("creating error stream for port forward: %w", err)
+	}
+	// we're not writing to this stream, but it is required for other streams to be able to connect
+	errorStream.Close()
+
+	errorChan := make(chan error)
+	go func() {
+		message, err := io.ReadAll(errorStream)
+		switch {
+		case err != nil:
+			log.Errorf("k8sPortFwd connection: reading from error stream: %v", err)
+		case len(message) > 0:
+			log.Errorf("k8sPortFwd connection: forwarding port: %v", string(message))
+		}
+		close(errorChan)
+	}()
```

I can't understand this code. errorChan is never used below, perhaps instead of printing the error it should be sent to the channel to be read and handled below? (or just remove the channel?)

---

Leftover code :) I removed the channel. There's no real place where we could send this error to - but the overlying protocol should also return an error, so I think we're fine.